### PR TITLE
3.0: Migrate EBS validators

### PR DESCRIPTION
**Pull request by enrico-usai**
## Migrate EBS volume type size validator
* Removed from mappings
* Tests were not present in the old code
* Improved unit tests to call the Validator and not the Schema.load().
  In this way the test will be more focused on the validation logic rather than calling other methods.

## Migrate EBS volume IOPS validator

* Removed from mappings
* Tests were not present in the old code
* Improved unit tests to call the Validator and not the Schema.load().
  In this way the test will be more focused on the validation logic rather than calling other methods.

## Migrate EBS volume throughput validator
* Removed from mappings.
* Tests were not present in the old code, now added.
* The old ebs_volume_throughput_validator has been splitted into two classes:
  EbsVolumeThroughputValidator and EbsVolumeThroughputIopsValidator
* Improved unit tests to call the Validator and not the Schema.load().
  In this way the test will be more focused on the validation logic rather than calling other methods.

## Migrate EBS volume size snapshot validator
* Removed from mappings.
* Migrated old tests. Removed the case on which the volume_size was None.

### CODE REVIEW

File: cli/src/pcluster/validators/common.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def __init__(self, raise_on_error=False):
     def _fail(self, message, level):
         raise ConfigValidationError
 
-    def _add_failure(self, message, level, fields=()):
+    def _add_failure(self, message: str, level: FailureLevel, fields=()):
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/validators/common.py:56]: minor: why not also add type hint for fields? For example: `fields:List[Param]=()`

File: cli/tests/pcluster/validators/test_ebs.py [hunk 2]
```diff
@@ -10,116 +10,194 @@
...
-            {"VolumeType": "gp3", "Throughput": 760, "Iops": 3000},
-            "Throughput to IOPS ratio of .* is too high",
-        ),
-        ({"VolumeType": "gp3", "Throughput": 760, "Iops": 10000}, None),
-        (
-            {"VolumeType": "gp3", "Throughput": 1001, "Iops": 2900},
-            [
-                "IOPS rate must be between 3000 and 16000 when provisioning gp3 volumes.",
-                "Throughput must be between 125 MB/s and 1000 MB/s when provisioning gp3 volumes.",
-            ],
```
Comment (hanwen-pcluste) [inline cli/tests/pcluster/validators/test_ebs.py:40]: Are we going to test multiple failure messages? My concern was that now we are trying to report multiple failures at once. By testing validators individually,  can we make sure these wrong inputs do not trigger any exceptions during the execution of the chain of the validators? If not, are we going to test it somewhere else?

**Review by hanwen-pcluste — APPROVED** [review]

File: cli/src/pcluster/validators/common.py (hunk 1 above)
Comment (enrico-usai) [reply cli/src/pcluster/validators/common.py:56]: Fixed, thanks

File: cli/tests/pcluster/validators/test_ebs.py (hunk 2 above)
Comment (enrico-usai) [reply cli/tests/pcluster/validators/test_ebs.py:40]: For multiple failures from the same validator we should extend these unit tests.
For multiple failures not related to the same validator we can create more unit tests loading the section as before.
The test for the "chain of validators" should be validated by the unit tests related to the priority feature.

File: cli/src/pcluster/validators/common.py [hunk 3]
```diff
@@ -53,20 +53,21 @@ def __init__(self, raise_on_error=False):
     def _fail(self, message, level):
         raise ConfigValidationError
 
-    def _add_failure(self, message: str, level: FailureLevel, fields=()):
+    def _add_failure(self, message: str, level: FailureLevel, failed_params: List[Param] = None):
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/validators/common.py:56]: `List[Param]=()` works, doesn't it? 
I am curious why you have to add another line of code for `if failed_params:`

Comment (enrico-usai) [reply cli/src/pcluster/validators/common.py]: Fixed

**Pull request merged**